### DISCUSSION TITLE
Stratum-varying Betas, testing

### DISCUSSION THREAD

**Discussion by lzachmann** [lzachmann/models-for-missing-data#8] [General]
Hey @eborgman thought I'd try this out via Discussions, in part to be a bit more disciplined about issues, and keeping issue content focused on the basic problem at hand (math / code, bugs, etc., as opposed to communications and scheduling). Let me know how this works! 

Let's do tackle this promptly this week! I'm around between 2pm and 4pm tomorrow (MT), and any time after noon on Wed. If there's a window that works for you just shoot me an invite. I think I resolved that problem today related to whether we ought to scale (or not) the Xs in each strata prior to fitting. I've also been working feverishly on some content for Sonya so I'm not the logjam there, too. Excited to show you all the progress.

Let's wait on rebooting your old Mac unless it's super needed. It may be easier to spin up the VM if need be.

**Comment by lzachmann** [comment]
Also Erin, any chance you could try to rerun the model that was throwing that error, the one with 

```yml
additional covariates:
    - Botanist (JA), deficit.pregr, deficit.pregr*MDCATY
```
and copy the resulting error message? That may help me with a quick Google Search. Things seem to be working over here on my end, which is probably good _and_ bad news.

**Reply by eborgman** [reply]
This error stinks of character limits on Windows: the "cannot open the connection" and the fact that it's running fine on your computer are fishy.

Original error
![image](https://user-images.githubusercontent.com/25830271/138783143-725466f5-b052-4360-9ef1-5ced248e06d5.png)

Traceback
![image](https://user-images.githubusercontent.com/25830271/138785538-2729f9c4-d76e-4021-8fc3-4585fbee7b22.png)

Retried limiting characters as much as I could from the config
![image](https://user-images.githubusercontent.com/25830271/138787262-e6f96338-097d-48d2-a24c-f390e2a6a492.png)

I tried overriding the limit on my machine, but I've lost administrative privileges, so I can't at the moment. Are there other places I could cut down on characters?

**Reply by lzachmann** [reply]
I can definitely cut down on characters. Will give that a shot in the morning first thing. I think your intuition here is probably right

**Comment by lzachmann** [comment]
Hey Erin I made a couple of changes. We can test tomorrow when you are available. Briefly, I went with the response column name in place of the response description, which should shorten things up a bit, and I replaced asterisks with X in interaction terms. I would bet a couple of dollars it was the latter, not the total number of characters, that was the culprit. 

We'll need to use git bash to get into the project directory, and then run
```sh
cd model-api
git pull origin gh-submodule
```
to get the latest.

**Reply by eborgman** [reply]
We're in business! I bet you're right - the issues was the asterisks. This new version is running. I had to manually install `hrbrthemes` and `mapview` to plot anything. It installed `phantomjs` by itself when it was running. 

As you probably know, none of the site/strata/park/marginal effects folders have anything in them - it doesn't know how to plot interactions I'm guessing.

It is not looking likely I'll make our 3:00. I'll have 3-3:15 free and after 3:45 available if we need to touch base.

**Reply by eborgman** [reply]
Quick update - I was also able to run this version with GRKO Mgmt Zone's with a deficit*MDCATY term and it worked! 

Along the way I discovered no figures are printing in the site/strata/park/me folders whether there's an interaction term or not. Trace and funnel plots print, as do ppc and coef plots.

**Reply by lzachmann** [reply]
Hey Erin, no problems on meeting today. Maybe we can catch up tomorrow? That's great news on GRKO, though I'd be interested to hear if perhaps we would need the Betas to vary by management zone, not strata?

**Reply by eborgman** [reply]
There's a close relationship between strata and mgmt zone: 2 strata are mostly irrigated, 2 strata are mostly upland non-irrigated, and 1 strata is riparian. It's not perfect though and there are some sites in the strata that are in a different mgmt zone. I'll take a look at the output and see if it makes sense.

**Reply by lzachmann** [reply]
Ok, we could also attempt a deficit*<management zone> interaction, but not sure if we'll rediscover some identifiability issues. May be worth a shot either way.

**Reply by eborgman** [reply]
To see what would happen I tried deficit*MgmtZone and got an error:
`Error: Can't subset columns that don't exist.
x The column `MgmtZone` doesn't exist.`

This is what the data look like. There is a MgmtZone column ... is the issue that it's categorical?
![image](https://user-images.githubusercontent.com/25830271/138975614-ab2922b0-e1ae-43a5-8a3c-ed01226bcf4f.png)

**Reply by eborgman** [reply]
With further testing the deficit*MDCATY for GRKO models, everything's running well for most of the configs. The species-richness model is not running though. When trying to run, a browser pops up. If I press "c" there's several more browsers and then the model just stops. I did a search and the uncommented browsers are in `fit.r` (3 of them), `inference-stratum-scale` (one), and `serializers-slim` (two). The only other uncommented browser is associated with an if else in `get-density`. Do you know why this is popping up for GRKO richness and not LIBI richness or GRKO other models?

**Reply by eborgman** [reply]
disregard my last about GRKO rich not working. Pretty sure this is user error; I must have messed something up in config or something - a fresh copy of everything is running fine.

**Reply by eborgman** [reply]
Hi Luke. I recreated the issue with GRKO richness - it's the negative binomial model that's not working. A bunch of browsers come up and then the model stops running. The 0-inflated neg binomial model works fine. Do you know what the issue might be?

**Reply by lzachmann** [reply]
Got it. I will try to take a look later today. Sorry for the delays -- stuck in meetings :(

**Reply by lzachmann** [reply]
Hey Erin I have a couple of ideas as to what may be happening here. I have other obligations here this afternoon, but will try to take a closer look tonight and tomorrow morning. Basically I think it has to do with issues created by 'hidden state' -- objects in the global environment that are overriding or conflicting with scoped environments. I wrote much of the pipeline before I was aware of these sorts of issues. Now it's time to deal with my past missteps. I'll keep updates posted here: #12. 

I don't know why it's an issue now where it wasn't before, but it may have something to do with the update form R 3.x to R 4.x. Whatever the cause,  I'll find a solution that will be more appropriate going forward.

**Reply by lzachmann** [reply]
Sorry I should have said the reason I noticed this was that I was seeing LIBI info (e.g., 3 strata total) show up when I was trying to run a GRKO model (which has 5 strata). Will know more soon, thanks for your patience.

**Comment by lzachmann** [comment]
Hey Erin you may want to check the args starting near line 50. I set `override_dq` (override derived quantities) to `TRUE` for whatever reason (testing, probably). Maybe try it again but set `override_dq` to `FALSE`?

```R
  n_adapt <- 2500
  n_update <- 5000
  n_iter <- 5000
  plot_bgcolor <- 'default'
  n_cores <- 1
  submit_as_test <- FALSE
  prep_for_hpc <- FALSE
  excl_null_mods <- FALSE
  override_dq <- T#FALSE
  bad_logs <- "none"
```

**Reply by eborgman** [reply]
yep, that did it!

**Comment by eborgman** [comment]
*Is_Irrigated interaction*

Here are some results with the following covariates for GRKO: Botanist, MgmtZone (model doesn't run without it), deficit, Is_Irrigated, deficit*Is_Irrigated.

What is confusing me is whether I need to include Is_Irrigated by itself AND in the interaction term and, if so, how to interpret the offset with both deficit and Is_Irrigated by themselves.

![additional-coef-estimates-Beta-untransformed](https://user-images.githubusercontent.com/25830271/139155929-2b1df416-f3e9-4d40-9fa4-ded85ce539e0.png)

**Reply by eborgman** [reply]
@megcswan @cllmcintyre @clivensperger @lzachmann If anyone has some brainpower to spare this Friday, can you help me think through something? After our call last Friday and further talking with Luke, I've landed on an interaction term of water deficit * IsIrrigated. This should account for the fact that irrigated areas of the park should not respond to the water balance term the same way as unirrigated areas of the park.

In the model, I've specified deficit, IsIrrigated, and deficit*IsIrrigated as covariates (along with botanist and mgmt zone, which were already in the model). I forget what the response was - let's say native richness. The results are in the post above. If I'm interpreting this right (and this is where you come in - please help!), the deficit histogram (histogram 2) is showing a positive relationship between deficit and native richness in unirrigated areas (IsIrrigated = 0), meaning more native species seen when it's drier. The deficit*IsIrrigated is an offset from histogram 2, representing the relationship of deficit in irrigated areas (IsIrrigated = 1) and native richness. It's slightly positive, which, when added to histogram 2, would create a less strong relationship between deficit and richness in irrigated areas. 

Does the IsIrrigated histogram (IsIrrigated term by itself, histogram 4) play into the offset of deficit?

(For my own notes - I think the IsIrrigated histogram is saying that irrigated areas (irrigated=1) have less species richness compared to the default unirrigated areas, which makes sense.)

I'd appreciate any insight you have in interpreting these interaction offsets.

**Reply by lzachmann** [reply]
Hey Erin I'm not totally sure that we can talk about them as being offsets exactly. Here's some code we should talk about when we get a chance. So far as I recall, you can't really directly interpret main effects if their interaction effect is 'significant':

```R
library(tidyverse)

# Covariates.
deficit <- seq(-2, 2, 0.1) # deficit plus or minus 2 standard deviations
is_irrigated <- c(0, 1) # irrigation: no (0) or yes (1)

# Coefficients.
intercept <- 2
b_deficit <- 0.25
b_is_irrigated <- -0.4
b_intx <- 0.15 # probably a bit exaggerated

d <- crossing(deficit, is_irrigated) %>% 
  mutate(
    # Compute the interaction term.
    # Compare to: model.matrix(~deficit*is_irrigated).,
    `deficit:is_irrigated` = deficit * is_irrigated,
    mu = exp(intercept +
               b_deficit * deficit +
               b_is_irrigated * is_irrigated +
               b_intx * `deficit:is_irrigated`))

p1 <- ggplot(d) +
  geom_line(aes(x = deficit, y = mu, color = factor(is_irrigated))) +
  theme(legend.position = 'bottom')
p2 <- ggplot(d %>% filter(deficit %% 1 == 0)) +
  geom_line(aes(x = is_irrigated, y = mu, color = factor(deficit)))+
  theme(legend.position = 'bottom') +
  scale_x_continuous(breaks = c(0, 1), labels = c('Unirrigated', 'Irrigated'))
cowplot::plot_grid(p1, p2, nrow = 1)

# ---- Notes ------------------------------------------------------------------

# Equation with line numbers (referred to below).
# (L1)  intercept +
# (L2)    b_deficit * deficit +
# (L3)    b_is_irrigated * is_irrigated + 
# (L4)    b_intx * is_irrigated:deficit

# b0 + (b1 * x1) + (b2 * x2) + (b3 * x1 * x2)

# Let's take the right-most panel first...
# - Mean native species richness is lower at irrigated sites than at non-
#   irrigated sites. The magnitude of the difference between irrigated and 
#   non-irrigated sites, however, tends to dissipate as deficit increases.

# Now, the left panel...
# - In general native species richness is higher when deficit is highest.
# - When deficit is at its average (i.e., when deficit == 0), *both* L1 and L4 
#   drop out of the linear predictor, see d %>% filter(deficit == 0). In other
#   words, the difference we see reduces to the effect of irrigation. At any
#   other value of deficit, the intX term is in play. This makes things 
#   interesting.
```

**Reply by lzachmann** [reply]
![3874760d-f274-41f7-a53c-d2c0395009d8](https://user-images.githubusercontent.com/7574887/139508699-94ff1ce5-1ded-4aac-ae64-9e9d2e72e47a.png)

**Reply by eborgman** [reply]
Thanks, Luke. I was conflating our past discussion about offsets related to Betas varying by stratum with this scenario. This is just standard interaction interpretation. Thanks for setting me back on the right path! Your code is super helpful.

**Comment by lzachmann** [comment]
Hey Erin I wanted to provide some thoughts in response to your question / concern above before it escaped me.

> What is confusing me is whether I need to include Is_Irrigated by itself AND in the interaction term and, if so, how to interpret the offset with both deficit and Is_Irrigated by themselves

I think the answer is that we _can_ drop one or both main effects, but perhaps ought not. I'm including my notes below in case they are of any interest. 

---

Consider a model involving a categorical (`is_irrigated`) by continuous (`b_deficit`) interaction term:

```
intercept + b_is_irrigated * is_irrigated + (b_deficit + b_deficit_offset * is_irrigated) * deficit                # (eq 1)
intercept + b_is_irrigated * is_irrigated + b_deficit * deficit + b_deficit_offset * is_irrigated * deficit    # (eq 1, rearranged)
```

Are both main effects &ndash; `b_is_irrigated * is_irrigated` and `b_deficit * deficit` &ndash; really required when an effect involving their interaction (`b_deficit_offset * is_irrigated * deficit`) is present? 

> The simple answer is no, you don’t always need main effects when there is an interaction. However, the interaction term will not have the same meaning as it would if both main effects were included in the model.
&mdash; [IDRE Stats](https://stats.idre.ucla.edu/stata/faq/what-happens-if-you-omit-the-main-effect-in-a-regression-model-with-an-interaction/)

Following the basic rationale in [this explanation](https://statisticalanalysisconsulting.com/why-you-should-include-main-effects-when-you-include-an-interaction/), we include both main effects to ensure that the information they contain about the mean is available to the model in making predictions. Consider a model in which we have dropped the main effect of both `is_irrigated` and `deficit`; the model reduces to `alpha + beta * is_irrigated * deficit`. Any observation for which the product of  `is_irrigated` and `deficit` is the same would have the same mean. Let's assume site 1 is `(is_irrigated == 0) * (deficit == 2.5)` while site 2 is `(is_irrigated == 1) * (deficit == 0)`. Of course, in both of these cases the interaction term goes to zero and we're left with the intercept alone. This is very unlikely to be a great model IF we know that irrigation and water deficit matter.  If unirrigated sites have generally higher native species richness, and higher water deficits factor native species, then site 1 really ought to have a higher mean than site 2. But the current, reduced model is incapable of producing that type of prediction.

**Reply by eborgman** [reply]
Getting some thoughts in response down...

@lzachmann do you have a sense whether it's best to 1) keep the main effect of 'is irrigated' in the model but have the model overestimate sigma for some strata (poisson model) or 2) use a model that fits the data better (neg binomial) but omit 'is irrigated' as a main effect?

In trying to figure out this interaction between deficit and irrigation, I've run several models. I started with the poisson, which did converge. It doesn't accurately capture the variation in the data though, so I also tried a negative binomial. This model wouldn't converge with 'is irrigated' as a main effect, so I tried omitting it. Here are summaries of what I found:

**poisson model (overestimates sd - or is it underestimate? - but otherwise converges)**
p_sd,p_mean,ppl,dic,gelman_diag
0.01,0.51,10476,4667,1.01

response: native species richness
covariates: botanist, mgmtzone, topography, deficit, is irrigated, deficit*is irrigated
interesting this one converges with 'is irrigated' as main effect

![image](https://user-images.githubusercontent.com/25830271/142936381-65f3c2b1-3d01-435a-890c-1cf0f31a3610.png)


![image](https://user-images.githubusercontent.com/25830271/142934763-84818f86-e4bc-4298-8d1e-9967a42e70ab.png)

**negative binomial (vanilla) model**
this one captures the data's variation much better
p_sd,p_mean,ppl,dic,gelman_diag
0.5,0.49,12238,4568,1.06

response: native species richness
covariates: botanist, mgmtzone, topography, deficit, deficit*is irrigated

this one does NOT converge with is irrigated as main effect, so it's omitted

![image](https://user-images.githubusercontent.com/25830271/142935426-59cb2a47-adf3-4dbe-bc1a-5a96272ab951.png)

**plots**
Using your code from earlier in the discussion, I get a pretty similar bottom line (but not completely), whether 'is irrigated' is in the model as main effect (poisson) or not (neg binomial)

poisson
![image](https://user-images.githubusercontent.com/25830271/142935635-db2f818b-c42b-434b-8d5e-3ccaa31e5c70.png)


neg binomial
![image](https://user-images.githubusercontent.com/25830271/142935572-46b89dc3-5529-4578-9a8a-b42a2dc79bff.png)

to create this figure, I just commented out the coefficient for the 'is irrigated' main effect.
I'm also ignoring the topography effect for now

#new model without irrigation main effect
d <- crossing(deficit, is_irrigated) %>% 
  mutate(
    `deficit:is_irrigated` = deficit * is_irrigated,
         mu = exp(intercept +
                    b_deficit * deficit +
                    #b_is_irrigated * is_irrigated +
                    b_intx * `deficit:is_irrigated`))

**Reply by eborgman** [reply]
And another thought related to your last post, thinking about how the model is specified

B0 + (B1 * deficit) + (B2 * deficit * is irrigated)

When 'is irrigated' is 0, that term drops out and we're left with the deficit term only (plus intercept). When 'is irrigated' is 1, the last term in that equation will basically add the deficit term again to the model. That doesn't seem right.

**Reply by lzachmann** [reply]
Hey Erin, thanks for such a thorough post. Lots to chew on here. A few initial thoughts, and we may want to follow up by Google Meet tomorrow or Wed if you're around. First I'm wondering if the reparameterized negative binomial might work even if the plain vanilla nb did not. I just need to wrap that up so it's functional with the finite population correction. We could try that when we chat. (It may be the fixed stratum-level kappa term is not giving the model the flexibility it needs to converge, whereas the newer nb model offers more flexibility.)

We may want to ensure that `deficit` in the data frame you are creating is 'scaled' to zero mean and unit variance. Probably it's just a sequence of numbers from -2 to 2 or so, so that's probably fine as is. Flagging just in case we need to revisit. We probably ought to look at the JAGS data object to see what the actual range of the scaled deficit values is. It may run from close to 0 to 5 or some such if it was really skewed to begin with, so evaluating deficit over a range -2 to 2 may not be all that sensible. Worth a look. 

Your last question about the model specification is a good one. I'm not seeing anything outright wrong with it, but will give this some additional thought. Note that when is_irrigated is 1, deficit does indeed appear again in the model, but this time associated with an entirely different coefficient (B2), which is allowed to be zero if there's no effect.... I'd love to think through the underlying ecology of these interaction terms too to be sure they pass the straight face test. If they don't, we've got a surprising new result _or_ (more likely) a model that is pathologically misspecified for reasons we don't yet understand.

**Reply by eborgman** [reply]
I'd love to chat Tuesday if you're around. Any chance you're available at about 11:30? If not, I could meet around 1 or 2 (or possibly 10). Feel free to send a google invite!

**Reply by eborgman** [reply]
In case you're dying to see these histograms, here they are:

native richness histograms faceted by management zone
![image](https://user-images.githubusercontent.com/25830271/143153315-f738014e-2fca-47b8-925c-93ec780d8771.png)

same faceted by not irrigated (0) and irrigated (1)
![image](https://user-images.githubusercontent.com/25830271/143153325-9ba2c9c3-2918-4412-9523-9baa11c62528.png)

**Reply by lzachmann** [reply]
Thanks Erin, the "deflections" (coefficients) for each management zone align nicely with the empirical distributions above.

<p align="center">
  <img src="https://user-images.githubusercontent.com/7574887/143291431-78b8931e-c6a3-4019-908b-78fbac57a64f.png" width="500">
</p>

I'm not sure I'm making a tremendous amount of headway just yet, but here's another way to think about this model. The following plot shows the mean in each management zone (facet / panel) and stratum (individual lines within each panel) over the empirical range of deficit values in each zone and stratum, respectively. (Note that this is for the Poisson model, which we know fails some of the posterior predictive checks, but which seems to do an okay job otherwise.) I fit the model once with the deficit-irrigation interaction term, and again without. See the results below. 

| **deficit + deficit*IsIrr** (no IsIrr main effect)   |  **deficit main effect only** (no intx term) |
|---|---|
|  ![ac594990-c2f3-4bca-998f-9d10ee3f900c](https://user-images.githubusercontent.com/7574887/143303390-07a93c53-c5b8-41ee-8a3e-a5a4f175698f.png) |  ![c2e0ac44-6b9c-4119-814d-14b8fcaebe96](https://user-images.githubusercontent.com/7574887/143303462-5a628648-909b-4f41-82f1-1d5abac05652.png) |

Fairly intuitively, I think, the interaction term is simply allowing the influence of deficit to be different depending on whether a site happens to be irrigated or not. Thus, the influence of deficit is more pronounced in "IrrigatedCentral". If this feels like the sort of model that you'd want, I'd say you're probably on firm ground using it. (Algebraically, the coefficient for the deficit-irrigation interaction term amounts to an offset for the main effect of deficit when a site is irrigated. Assuming we leave the main effect of `IsIrr` out, `Beta_deficit * deficit + Beta_intx * deficit * IsIrr` is the same as `(Beta_deficit + Beta_intx * IsIrr) * deficit`.)

I'm attaching the kludgy code I was using to investigate this in case we need to revisit. `ex_path` refers to the model output directory in each case. 

```R
library(tidyverse)

coefs <- read_csv(file.path(
  ex_path, '03-inference', 'coda-samples-quantiles.csv'
))

mgmt_zone_lkp <- read_csv(file.path(
  ex_path, '00-input', 'encoding-for-MgmtZone.csv'
))

stratum_b0 <- coefs %>% 
  filter(grepl("^B0", param)) %>% 
  mutate(stratum_index = as.integer(
    str_extract(param, "(?<=\\[)\\d(?=\\])")
  )) %>% 
  select(b0 = `50%`, stratum_index)
other_fixed_coefs <- coefs %>%
  filter(!is.na(covariate)) %>% # grepl("^MgmtZone", covariate)
  select(covariate, `50%`) %>% 
  pivot_wider(names_from = 'covariate', values_from = '50%') %>% 
  rename_with(~paste("Beta", .x, sep = "_"), everything())
d_tmp <- d %>% 
  select(stratum_index, matches('fe_MgmtZone\\d')) %>% 
  distinct() %>% 
  left_join(stratum_b0, by = "stratum_index") %>% 
  bind_cols(other_fixed_coefs) %>% 
  left_join(mgmt_zone_lkp)

is_irr_lkp <- d %>% 
  left_join(mgmt_zone_lkp) %>% 
  group_by(stratum_id, stratum_index, MgmtZone) %>% 
  # mutate(fe_IsIrr = as.integer(grepl("^Irrigated", MgmtZone))) %>% 
  mutate(fe_IsIrr = as.integer(fe_dXIsIrr != 0)) %>%
  distinct(fe_IsIrr) %>% 

d_eval <- d %>% 
  left_join(mgmt_zone_lkp) %>% 
  left_join(is_irr_lkp) %>% 
  group_by(stratum_id, stratum_index, MgmtZone) %>% 
  expand(fe_d = seq(min(fe_d), max(fe_d), length.out = 20),
         fe_IsIrr = unique(fe_IsIrr)) %>% 
  mutate(is_min_or_max = fe_d %in% range(fe_d)) %>% 
  ungroup() %>% 
  left_join(d_tmp) %>% 
  mutate(fe_dXIsIrr = fe_d * fe_IsIrr,
         mu = exp(b0 +
                    Beta_MgmtZone1 * fe_MgmtZone1 +
                    Beta_MgmtZone2 * fe_MgmtZone2 +
                    Beta_MgmtZone3 * fe_MgmtZone3 +
                    Beta_MgmtZone4 * fe_MgmtZone4 +
                    Beta_d * fe_d +
                    Beta_dXIsIrr * fe_dXIsIrr))

ggplot(d_eval) +
  facet_wrap(~MgmtZone) +
  # facet_grid(MgmtZone~stratum_id) +
  geom_line(aes(x = fe_d, y = mu, color = factor(fe_IsIrr), group = stratum_id)) +
  theme(legend.position = 'bottom')
```

**Reply by eborgman** [reply]
Thanks, Luke. This is helpful. It does make sense to allow the influence of deficit to be different based on irrigation status. I'm a little surprised the 2 figures you showed (with interaction and without) are as similar as they are. (The y axes are slightly different.) I'll let this all sink in and think more about it. Thank you!
